### PR TITLE
add multinomial probability distribution

### PR DESCRIPTION
**Pull request by cxxly**
### PR types

New features
### PR changes

APIs
### Describe

- add multinomial distribution with mean, variance,sample,entropy,prob,log_prob method.
- update beta,dirichet,exponential family docs.
- fix categorical entropy,sample bugs.

### CODE REVIEW

File: python/paddle/distribution/kl.py [hunk 1]
```diff
@@ -68,6 +68,11 @@ def kl_divergence(p, q):
 def register_kl(cls_p, cls_q):
     """Decorator for register a KL divergence implemention function.
 
+    when call ``kl_divergence(p, q)`` , will search concrete implemention
```
Comment (iclementine) [inline python/paddle/distribution/kl.py]: 注意语法。

File: python/paddle/distribution/beta.py [hunk 2]
```diff
@@ -37,8 +44,15 @@ class Beta(ExponentialFamily):
 
 
     Args:
-        alpha (float|Tensor): alpha parameter of beta distribution, positive(>0).
+        alpha (float|Tensor): alpha parameter of beta distribution, 
+            positive(>0), support broadcast semantic. when the parameter is
```
Comment (iclementine) [inline python/paddle/distribution/beta.py]: 注意英文句法，首字母大写等。

File: python/paddle/distribution/categorical.py [hunk 3]
```diff
@@ -200,7 +208,8 @@ def kl_divergence(self, other):
         if not in_dygraph_mode():
             check_type(other, 'other', Categorical, 'kl_divergence')
 
-        logits = self.logits - nn.reduce_max(self.logits, dim=-1, keep_dim=True)
+        logits = self.logits - \
+            nn.reduce_max(self.logits, dim=-1, keep_dim=True)
```
Comment (iclementine) [inline python/paddle/distribution/categorical.py]: 这些推荐使用 paddle.max ，而不是用 nn 里面的函数。

File: python/paddle/distribution/dirichlet.py [hunk 4]
```diff
@@ -22,23 +22,37 @@
 
 class Dirichlet(ExponentialFamily):
     r"""
-    Dirichlet distribution with parameter concentration
+    Dirichlet distribution with parameter concentration.
```
Comment (iclementine) [inline python/paddle/distribution/dirichlet.py]: concentration 可以打个引号什么的，语义才比较明确。

File: python/paddle/distribution/dirichlet.py [hunk 5]
```diff
@@ -22,23 +22,37 @@
...
+
+        B(\boldsymbol \alpha) = \frac{\prod_{i=1}^{k} \Gamma(\alpha_i)}{\Gamma(\alpha_0)}
+
+    :math:`\alpha_0=\sum_{i=1}^{k} \alpha_i` is the sum of parameters, 
+    :math:`\Gamma(\alpha)` is gamma function.
 
     Args:
         concentration (Tensor): concentration parameter of dirichlet 
-            distribution
+            distribution, also called :math:`\alpha`. when concentration over
```
Comment (iclementine) [inline python/paddle/distribution/dirichlet.py]: 注意句子首字母大写。英文句子用英文逗号，且后附空格。

File: python/paddle/distribution/distribution.py [hunk 6]
```diff
@@ -102,7 +103,13 @@ def log_prob(self, value):
...
 
     def probs(self, value):
-        """Probability density/mass function."""
+        """Probability density/mass function.
+        
+        .. note:: 
+        
+            This method will be deprecated in the future, please use `prob` 
+            instead.
+        """
```
Comment (iclementine) [inline python/paddle/distribution/distribution.py:112]: 看看是否可以使用框架的 deprecated 装饰器。python/paddle/utils/deprecated.py

File: python/paddle/distribution/multinomial.py [hunk 7]
```diff
@@ -0,0 +1,184 @@
...
+        f(x_1, ..., x_k; n, p_1,...,p_k) = \frac{n!}{x_1!...x_k!}p_1^{x_1}...p_k^{x_k}
+
+    where, :math:`n` is number of trails, k is the number of categories, 
+    :math:`p_i` denote probability of a trail falling into each category, 
+    :math:`{\textstyle \sum_{i=1}^{k}p_i=1}, p_i \ge 0`, and :math:`x_i` denote 
+    count of each category. 
+
+    Args:
+        total_count (int): Number of trials.
+        probs (Tensor): Probability of a trail falling into each category. Last
```
Comment (iclementine) [inline python/paddle/distribution/multinomial.py]: 注意拼写。 trial trail

File: python/paddle/fluid/tests/unittests/distribution/test_distribution_multinomial.py [hunk 8]
```diff
@@ -0,0 +1,130 @@
...
+            tuple(samples.shape),
+            sample_shape + self._dist.batch_shape + self._dist.event_shape)
+        self.assertTrue(
+            np.all(samples.sum(-1).numpy() == self._dist.total_count))
+
+        sample_shape = (100000, )
+        samples = self._dist.sample(sample_shape)
+        sample_mean = samples.mean(axis=0)
+        np.testing.assert_allclose(
+            sample_mean, self._dist.mean, atol=0, rtol=0.20)
```
Comment (iclementine) [inline python/paddle/fluid/tests/unittests/distribution/test_distribution_multinomial.py:86]: 这样子的 tolerance 可能会被 CI 系统认为不合理。是否写明一下这么做的原因。

File: python/paddle/fluid/tests/unittests/distribution/test_distribution_multinomial.py [hunk 9]
```diff
@@ -0,0 +1,130 @@
...
+
+@config.place(config.DEVICES)
+@config.parameterize(
+    (config.TEST_CASE_NAME, 'total_count', 'probs', 'value'),
+    [
+        ('value-float', 10, np.array([0.2, 0.3, 0.5]), np.array([2., 3., 5.])),
+        ('value-int', 10, np.array([0.2, 0.3, 0.5]), np.array([2, 3, 5])),
+        ('value-multi-dim', 10, np.array([[0.3, 0.7], [0.5, 0.5]]),
+         np.array([[4., 6], [8, 2]])),
+        # ('value-sum-non-n', 10, np.array([0.5, 0.2, 0.3]), np.array([4,5,2])),
```
Comment (iclementine) [inline python/paddle/fluid/tests/unittests/distribution/test_distribution_multinomial.py:109]: 是否添加采多个样的 case. 比如 Batch shape 是 (), 而 sample shape 是 (2,） 类似这样的。

File: python/paddle/fluid/tests/unittests/distribution/test_distribution_multinomial_static.py [hunk 10]
```diff
@@ -0,0 +1,155 @@
...
+@config.parameterize((config.TEST_CASE_NAME, 'total_count', 'probs'), [
+    ('total_count_le_one', 0, np.array([0.3, 0.7])),
+    ('total_count_float', np.array([0.3, 0.7])),
+    ('probs_zero_dim', np.array(0)),
+])
+class TestMultinomialException(unittest.TestCase):
+    def setUp(self):
+        self.prog = paddle.static.Program()
+        self.exe = paddle.static.Executor()
+        with paddle.static.program_guard(prog):
```
Comment (iclementine) [inline python/paddle/fluid/tests/unittests/distribution/test_distribution_multinomial_static.py]: 最好写两个 program, 虽然这里并不创建参数。

File: python/paddle/distribution/distribution.py (hunk 6 above)
Comment (cxxly) [reply python/paddle/distribution/distribution.py:112]: 功能废弃需要走评审流程，先临时加个注释提醒用户

File: python/paddle/distribution/categorical.py (hunk 3 above)
Comment (cxxly) [reply python/paddle/distribution/categorical.py]: 这是之前同学遗留代码，我先更新下这部分代码，其余遗留在后续规划中，会统一更新

File: python/paddle/distribution/kl.py (hunk 1 above)
Comment (cxxly) [reply python/paddle/distribution/kl.py]: updated

File: python/paddle/distribution/beta.py (hunk 2 above)
Comment (cxxly) [reply python/paddle/distribution/beta.py]: updated

File: python/paddle/distribution/dirichlet.py (hunk 4 above)
Comment (cxxly) [reply python/paddle/distribution/dirichlet.py]: updated

File: python/paddle/distribution/multinomial.py (hunk 7 above)
Comment (cxxly) [reply python/paddle/distribution/multinomial.py]: updated

File: python/paddle/fluid/tests/unittests/distribution/test_distribution_multinomial.py (hunk 8 above)
Comment (cxxly) [reply python/paddle/fluid/tests/unittests/distribution/test_distribution_multinomial.py:86]: updated

File: python/paddle/distribution/kl.py [hunk 11]
```diff
@@ -68,9 +65,15 @@ def kl_divergence(p, q):
 def register_kl(cls_p, cls_q):
     """Decorator for register a KL divergence implemention function.
 
+    The ``kl_divergence(p, q)`` function will search concrete implemention 
+    function registered by ``register_kl``, according to multi-dispatch pattern. 
+    If find the implemention function, it will return the result, or not will 
+    raise ``NotImplementError`` exception. User can register implemention 
+    funciton by the decorator.
```
Comment (iclementine) [inline python/paddle/distribution/kl.py]: 1. implemention functions (plural forum);
2. If an implemention function is found;
3. ortherwise, it will raise a ``NotImplementError`` exception;
4. Users
5. functions

File: python/paddle/distribution/kl.py [hunk 12]
```diff
@@ -167,7 +170,7 @@ def _kl_uniform_uniform(p, q):
 
 @register_kl(ExponentialFamily, ExponentialFamily)
 def _kl_expfamily_expfamily(p, q):
-    """compute kl-divergence using `Bregman divergences` 
+    """Compute kl-divergence using `Bregman divergences` 
     https://www.lix.polytechnique.fr/~nielsen/EntropyEF-ICIP2010.pdf
```
Comment (iclementine) [inline python/paddle/distribution/kl.py]: Use correct hyperlink format of rst.

File: python/paddle/distribution/kl.py [hunk 13]
```diff
@@ -205,5 +208,5 @@ def _kl_expfamily_expfamily(p, q):
 
 
 def _sum_rightmost(value, n):
-    """sum value along rightmost n dim"""
+    """Sum value along rightmost n dim"""
```
Comment (iclementine) [inline python/paddle/distribution/kl.py]: elements ...(plural form)
dimensions.

File: python/paddle/distribution/beta.py [hunk 14]
```diff
@@ -37,8 +44,14 @@ class Beta(ExponentialFamily):
 
 
     Args:
-        alpha (float|Tensor): alpha parameter of beta distribution, positive(>0).
-        beta (float|Tensor): beta parameter of beta distribution, positive(>0).
+        alpha (float|Tensor): Alpha parameter. It support broadcast semantic.
```
Comment (iclementine) [inline python/paddle/distribution/beta.py]: It supports
Sementics.
is a tensor
represents
distributions
a

File: python/paddle/distribution/dirichlet.py [hunk 15]
```diff
@@ -22,23 +22,37 @@
...
+        B(\boldsymbol \alpha) = \frac{\prod_{i=1}^{k} \Gamma(\alpha_i)}{\Gamma(\alpha_0)}
+
+    :math:`\alpha_0=\sum_{i=1}^{k} \alpha_i` is the sum of parameters, 
+    :math:`\Gamma(\alpha)` is gamma function.
 
     Args:
-        concentration (Tensor): concentration parameter of dirichlet 
-            distribution
+        concentration (Tensor): "Concentration" parameter of dirichlet 
+            distribution, also called :math:`\alpha`. When It's over one
```
Comment (iclementine) [inline python/paddle/distribution/dirichlet.py]: When it's.

File: python/paddle/distribution/dirichlet.py [hunk 16]
```diff
@@ -22,23 +22,37 @@
...
+    :math:`\alpha_0=\sum_{i=1}^{k} \alpha_i` is the sum of parameters, 
+    :math:`\Gamma(\alpha)` is gamma function.
 
     Args:
-        concentration (Tensor): concentration parameter of dirichlet 
-            distribution
+        concentration (Tensor): "Concentration" parameter of dirichlet 
+            distribution, also called :math:`\alpha`. When It's over one 
+            dimension, the last axis is parameter of distribution,
+            ``event_shape=concentration.shape[-1:]`` , other axes is batch
```
Comment (iclementine) [inline python/paddle/distribution/dirichlet.py]: are

Comment (iclementine) [reply python/paddle/distribution/dirichlet.py]: axes other than the last are condsiderd batch dimensions.

File: python/paddle/distribution/multinomial.py [hunk 17]
```diff
@@ -0,0 +1,184 @@
...
+
+    where, :math:`n` is number of trials, k is the number of categories, 
+    :math:`p_i` denote probability of a trial falling into each category, 
+    :math:`{\textstyle \sum_{i=1}^{k}p_i=1}, p_i \ge 0`, and :math:`x_i` denote 
+    count of each category. 
+
+    Args:
+        total_count (int): Number of trials.
+        probs (Tensor): Probability of a trial falling into each category. Last 
+            axis of probs indexes over categories, other axes index over batches.
```
Comment (iclementine) [inline python/paddle/distribution/multinomial.py:47]: The last axis

**Review by iclementine — DISMISSED** [review]
LGTM

TODO: refine documentation in next PR!

**Review by iclementine — APPROVED** [review]

**Review by dingjiaweiww — APPROVED** [review]

**Review by jeff41404 — APPROVED** [review]
approve

**Review by XiaoguangHu01 — APPROVED** [review]
LGTM
